targets/TARGET_STM/i2c_api.c

```diff
@@ -258,7 +265,77 @@ void i2c_sw_reset(i2c_t *obj)
     handle->Instance->CR1 |=  I2C_CR1_PE;
 }
 
-void i2c_init(i2c_t *obj, PinName sda, PinName scl)
+#ifdef DEVICE_I2CSLAVE
+
+static int i2c_slave_read(i2c_t *obj, char *data, int length)
+{
+    struct i2c_s *obj_s = I2C_S(obj);
+    I2C_HandleTypeDef *handle = &(obj_s->handle);
+    int count = 0;
+    int ret = 0;
+    uint32_t timeout = 0;
+
+    /*  Always use I2C_NEXT_FRAME as slave will just adapt to master requests */
+    ret = HAL_I2C_Slave_Sequential_Receive_IT(handle, (uint8_t *) data, length, I2C_NEXT_FRAME);
+
+    if (ret == HAL_OK) {
+        timeout = BYTE_TIMEOUT_US * (length + 1);
+        while (obj_s->pending_slave_rx_maxter_tx && (--timeout != 0)) {
+            wait_us(1);
+        }
+
+        if (timeout != 0) {
+            count = length;
+        } else {
+            DEBUG_PRINTF("TIMEOUT or error in i2c_slave_read\r\n");
+        }
+    }
+    return count;
```

Comment:
Shouldn't be all return values based on 'handle->XferCount' to reflect real sent/received count